tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/variable/Variable.java

```diff
@@ -0,0 +1,414 @@
+/*
+  Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ ==============================================================================
+ */
+package org.tensorflow.variable;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.function.Supplier;
+import org.tensorflow.ExecutionEnvironment;
+import org.tensorflow.Operand;
+import org.tensorflow.Operation;
+import org.tensorflow.Output;
+import org.tensorflow.ndarray.Shape;
+import org.tensorflow.op.Op;
+import org.tensorflow.op.Operands;
+import org.tensorflow.op.Scope;
+import org.tensorflow.op.annotation.Endpoint;
+import org.tensorflow.op.annotation.Operator;
+import org.tensorflow.op.core.AssignAddVariableOp;
+import org.tensorflow.op.core.AssignSubVariableOp;
+import org.tensorflow.op.core.AssignVariableOp;
+import org.tensorflow.op.core.Init;
+import org.tensorflow.op.core.IsVariableInitialized;
+import org.tensorflow.op.core.ReadVariableOp;
+import org.tensorflow.op.core.VarHandleOp;
+import org.tensorflow.proto.framework.DataType;
+import org.tensorflow.types.TBool;
+import org.tensorflow.types.family.TType;
+
+/**
+ * A class representing a read-only tensor variable with a constant shape and data type.  Analogous to Python's tf.Variable.
+ * Any access will always return the most recent assignment.
+ * <p>
+ * Supports eager and graph mode, and will use {@code VariableV2} in graph mode and enforce ordered assignments.
+ * <p>
+ * Provides methods to get the value and initialize the value if it hasn't already been set.
+ * Also implements {@code Operand<T>} using the stored value.
+ * <p>
+ * Variables will be registered in their execution environment's {@link ExecutionEnvironment#variables()}.
+ *
+ */
+@Operator
+public class Variable<T extends TType> implements Operand<T> {
+
+  private final Scope initialScope;
+  private final String name;
+
+  private final Shape shape;
+  private final DataType dataType;
+  private final boolean trainable;
```

Comment:
It's not in python, so I copied that.  It doesn't seem that hard to do, but we'd want to have something to ensure it's not changed while a gradient tape is active in eager mode, or have some special handling for that.  It depends on what that API looks like though.

I don't know how keras handles it.  Maybe with the StopGradient op?

Reply:
Ah ok, so the python docs make sense. The notes on the behaviour of trainable in saved model means it's still counter-intuitive as I think I'd like to store the epoch count in a checkpoint, it makes it easier to restart training, but oh well.

Reply:
It would be nice to add the trainable variables to a hook in the graph the same way they do in Python. The registerVariable hook could check if it's trainable.

Reply:
I could, yeah.  I figured it was trivial to filter the variables list.  I would think any optimizer implementations would want to take a list of variables from whatever types of models they use anyways, to allow for multiple models in the same graph.